pkg/ipam/crd.go

```diff
@@ -697,3 +697,34 @@ func (a *crdAllocator) RestoreFinished() {
 		close(a.store.restoreFinished)
 	})
 }
+
+// NewIPNotAvailableInPoolError returns an error resprenting the given IP not
+// being available in the IPAM pool.
+func NewIPNotAvailableInPoolError(ip net.IP) error {
+	return &ErrIPNotAvailableInPool{ip: ip}
+}
+
+// ErrIPNotAvailableInPool represents an error when an IP is not available in
+// the pool.
+type ErrIPNotAvailableInPool struct {
+	ip net.IP
+}
+
+func (e *ErrIPNotAvailableInPool) Error() string {
+	return fmt.Sprintf("IP %s is not available", e.ip.String())
+}
+
+// Is provides this error type with the logic for use with errors.Is.
+func (e *ErrIPNotAvailableInPool) Is(target error) bool {
+	if e == nil || target == nil {
+		return false
+	}
+	t, ok := target.(*ErrIPNotAvailableInPool)
+	if !ok {
+		return ok
+	}
+	if t == nil {
```

Comment:
Why is this check required after the type check @L722.

Reply:
It's possible for an interface to have a nil value so we need to check it after a successful type assertion.

Reply:
The first nil check for `t` might be superfluous, but not the check for `e`. I kept the check for `t` because it doesn't hurt.

Reply:
https://golang.org/ref/spec#Type_assertions

`asserts that x is not nil and that the value stored in x is of type T. The notation x.(T) is called a type assertion.`

Reply:
But we are asserting that `t` is `*ErrIPNotAvailableInPool`, which is a pointer and hence can be nil, no?